EntryListenerOnReconnectTest.testClusterReconnectMultipleServersDueToHeartbeatSmartRouting & testTemporaryBlockedNoDisconnectionMultipleServerSmartRouting

```
com.hazelcast.core.HazelcastException: Listener can not be added 
	at com.hazelcast.client.spi.impl.listener.ClientSmartListenerService$1.call(ClientSmartListenerService.java:81)
	at com.hazelcast.client.spi.impl.listener.ClientSmartListenerService$1.call(ClientSmartListenerService.java:67)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
	at java.lang.Thread.run(Thread.java:662)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
Caused by: java.io.IOException: Packet not send to [127.0.0.1]:5429
	at com.hazelcast.client.spi.impl.ClientInvocationServiceSupport.send(ClientInvocationServiceSupport.java:114)
	at com.hazelcast.client.spi.impl.ClientSmartInvocationServiceImpl.invokeOnConnection(ClientSmartInvocationServiceImpl.java:79)
	at com.hazelcast.client.spi.impl.ClientInvocation.invokeOnSelection(ClientInvocation.java:141)
	at com.hazelcast.client.spi.impl.ClientInvocation.invokeUrgent(ClientInvocation.java:134)
	at com.hazelcast.client.spi.impl.listener.ClientSmartListenerService.invoke(ClientSmartListenerService.java:112)
	at com.hazelcast.client.spi.impl.listener.ClientSmartListenerService.access$200(ClientSmartListenerService.java:47)
	at com.hazelcast.client.spi.impl.listener.ClientSmartListenerService$1.call(ClientSmartListenerService.java:77)
	at com.hazelcast.client.spi.impl.listener.ClientSmartListenerService$1.call(ClientSmartListenerService.java:67)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
	at java.lang.Thread.run(Thread.java:662)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:94)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:34)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:147)
	at com.hazelcast.client.spi.impl.listener.ClientSmartListenerService.invoke(ClientSmartListenerService.java:116)
	at com.hazelcast.client.spi.impl.listener.ClientSmartListenerService.access$200(ClientSmartListenerService.java:47)
	at com.hazelcast.client.spi.impl.listener.ClientSmartListenerService$1.call(ClientSmartListenerService.java:77)
	at com.hazelcast.client.spi.impl.listener.ClientSmartListenerService$1.call(ClientSmartListenerService.java:67)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
	at java.lang.Thread.run(Thread.java:662)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-OracleJDK1.6-freestyle/124/

Failed again in a PR builder:
```java
java.lang.AssertionError: CountDownLatch failed to complete within 120 seconds, count left: 1
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:973)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:966)
	at com.hazelcast.test.HazelcastTestSupport.assertOpenEventually(HazelcastTestSupport.java:958)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest.validateListenerFunctionality(AbstractListenersOnReconnectTest.java:510)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest.validateRegistrationsAndListenerFunctionality(AbstractListenersOnReconnectTest.java:465)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest.testListenersTerminateOwnerNode(AbstractListenersOnReconnectTest.java:451)
	at com.hazelcast.client.listeners.AbstractListenersOnReconnectTest.testListenersNonSmartRoutingMultipleServer(AbstractListenersOnReconnectTest.java:402)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:597)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:105)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:97)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)
	at java.lang.Thread.run(Thread.java:662)
```
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/9777/testReport/junit/com.hazelcast.client.listeners/EntryListenerOnReconnectTest/testListenersNonSmartRoutingMultipleServer/

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/AbstractListenersOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/AbstractListenersOnReconnectTest.java
@@ -33,6 +33,9 @@ import com.hazelcast.core.MemberAttributeEvent;
 import com.hazelcast.core.MembershipEvent;
 import com.hazelcast.core.MembershipListener;
 import com.hazelcast.nio.Connection;
+import com.hazelcast.spi.impl.NodeEngineImpl;
+import com.hazelcast.spi.impl.eventservice.impl.EventServiceImpl;
+import com.hazelcast.spi.impl.eventservice.impl.EventServiceSegment;
 import com.hazelcast.spi.properties.GroupProperty;
 import com.hazelcast.test.AssertTask;
 import org.junit.After;
@@ -49,6 +52,7 @@ import java.util.concurrent.atomic.AtomicInteger;
 
 import static com.hazelcast.client.spi.properties.ClientProperty.HEARTBEAT_TIMEOUT;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertTrue;
 
 public abstract class AbstractListenersOnReconnectTest extends ClientTestSupport {
@@ -461,11 +465,28 @@ public abstract class AbstractListenersOnReconnectTest extends ClientTestSupport
 
     private void validateRegistrationsAndListenerFunctionality() {
         assertClusterSizeEventually(clusterSize, client);
+        validateRegistrationsOnMembers(factory);
         validateRegistrations(clusterSize, registrationId, getHazelcastClientInstanceImpl(client));
         validateListenerFunctionality();
         assertTrue(removeListener(registrationId));
     }
 
+    protected void validateRegistrationsOnMembers(final TestHazelcastFactory factory) {
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                for (HazelcastInstance instance : factory.getAllHazelcastInstances()) {
+                    NodeEngineImpl nodeEngineImpl = getNodeEngineImpl(instance);
+                    EventServiceImpl eventService = (EventServiceImpl) nodeEngineImpl.getEventService();
+                    EventServiceSegment serviceSegment = eventService.getSegment(getServiceName(), false);
+                    assertNotNull(serviceSegment);
+                    assertEquals(1, serviceSegment.getRegistrationIdMap().size());
+                }
+            }
+        });
+    }
+
+    abstract String getServiceName();
 
     private void validateRegistrations(final int clusterSize, final String registrationId,
                                        final HazelcastClientInstanceImpl clientInstanceImpl) {

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/EntryListenerOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/EntryListenerOnReconnectTest.java
@@ -19,6 +19,7 @@ package com.hazelcast.client.listeners;
 import com.hazelcast.core.EntryAdapter;
 import com.hazelcast.core.EntryEvent;
 import com.hazelcast.core.IMap;
+import com.hazelcast.map.impl.MapService;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
@@ -30,6 +31,11 @@ import org.junit.runner.RunWith;
 public class EntryListenerOnReconnectTest extends AbstractListenersOnReconnectTest {
 
     private IMap<String, String> iMap;
+
+    @Override
+    String getServiceName() {
+        return MapService.SERVICE_NAME;
+    }
 
     @Override
     protected String addListener() {

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/ListItemListenerOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/ListItemListenerOnReconnectTest.java
@@ -16,6 +16,7 @@
 
 package com.hazelcast.client.listeners;
 
+import com.hazelcast.collection.impl.list.ListService;
 import com.hazelcast.core.IList;
 import com.hazelcast.core.ItemEvent;
 import com.hazelcast.core.ItemListener;
@@ -30,6 +31,11 @@ import org.junit.runner.RunWith;
 public class ListItemListenerOnReconnectTest extends AbstractListenersOnReconnectTest {
 
     private IList<String> iList;
+
+    @Override
+    String getServiceName() {
+        return ListService.SERVICE_NAME;
+    }
 
     @Override
     protected String addListener() {

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/ListenerTests.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/ListenerTests.java
@@ -24,7 +24,6 @@ import com.hazelcast.core.IMap;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
-import com.hazelcast.test.annotation.Repeat;
 import org.junit.After;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/MultiMapEntryListenerOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/MultiMapEntryListenerOnReconnectTest.java
@@ -19,6 +19,7 @@ package com.hazelcast.client.listeners;
 import com.hazelcast.core.EntryAdapter;
 import com.hazelcast.core.EntryEvent;
 import com.hazelcast.core.MultiMap;
+import com.hazelcast.multimap.impl.MultiMapService;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
@@ -30,6 +31,11 @@ import org.junit.runner.RunWith;
 public class MultiMapEntryListenerOnReconnectTest extends AbstractListenersOnReconnectTest {
 
     private MultiMap<String, String> multiMap;
+
+    @Override
+    String getServiceName() {
+        return MultiMapService.SERVICE_NAME;
+    }
 
     @Override
     protected String addListener() {

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/QueueItemListenerOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/QueueItemListenerOnReconnectTest.java
@@ -16,6 +16,7 @@
 
 package com.hazelcast.client.listeners;
 
+import com.hazelcast.collection.impl.queue.QueueService;
 import com.hazelcast.core.IQueue;
 import com.hazelcast.core.ItemEvent;
 import com.hazelcast.core.ItemListener;
@@ -30,6 +31,11 @@ import org.junit.runner.RunWith;
 public class QueueItemListenerOnReconnectTest extends AbstractListenersOnReconnectTest {
 
     private IQueue<String> iQueue;
+
+    @Override
+    String getServiceName() {
+        return QueueService.SERVICE_NAME;
+    }
 
     @Override
     protected String addListener() {

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/ReplicatedMapEntryListenerOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/ReplicatedMapEntryListenerOnReconnectTest.java
@@ -16,20 +16,34 @@
 
 package com.hazelcast.client.listeners;
 
+import com.hazelcast.client.test.TestHazelcastFactory;
 import com.hazelcast.core.EntryAdapter;
 import com.hazelcast.core.EntryEvent;
+import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.ReplicatedMap;
+import com.hazelcast.replicatedmap.impl.ReplicatedMapService;
+import com.hazelcast.spi.impl.NodeEngineImpl;
+import com.hazelcast.spi.impl.eventservice.impl.EventServiceImpl;
+import com.hazelcast.spi.impl.eventservice.impl.EventServiceSegment;
+import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
+import static org.junit.Assert.assertTrue;
+
 @RunWith(HazelcastParallelClassRunner.class)
 @Category({QuickTest.class, ParallelTest.class})
 public class ReplicatedMapEntryListenerOnReconnectTest extends AbstractListenersOnReconnectTest {
 
     private ReplicatedMap<String, String> replicatedMap;
+
+    @Override
+    String getServiceName() {
+        return ReplicatedMapService.SERVICE_NAME;
+    }
 
     @Override
     protected String addListener() {
@@ -51,5 +65,24 @@ public class ReplicatedMapEntryListenerOnReconnectTest extends AbstractListeners
     @Override
     public boolean removeListener(String registrationId) {
         return replicatedMap.removeEntryListener(registrationId);
+    }
+
+    @Override
+    protected void validateRegistrationsOnMembers(final TestHazelcastFactory factory) {
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                boolean found = false;
+                for (HazelcastInstance instance : factory.getAllHazelcastInstances()) {
+                    NodeEngineImpl nodeEngineImpl = getNodeEngineImpl(instance);
+                    EventServiceImpl eventService = (EventServiceImpl) nodeEngineImpl.getEventService();
+                    EventServiceSegment serviceSegment = eventService.getSegment(getServiceName(), false);
+                    if (serviceSegment != null && serviceSegment.getRegistrationIdMap().size() == 1) {
+                        found = true;
+                    }
+                }
+                assertTrue(found);
+            }
+        });
     }
 }

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/SetItemListenerOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/SetItemListenerOnReconnectTest.java
@@ -16,6 +16,7 @@
 
 package com.hazelcast.client.listeners;
 
+import com.hazelcast.collection.impl.set.SetService;
 import com.hazelcast.core.ISet;
 import com.hazelcast.core.ItemEvent;
 import com.hazelcast.core.ItemListener;
@@ -30,6 +31,11 @@ import org.junit.runner.RunWith;
 public class SetItemListenerOnReconnectTest extends AbstractListenersOnReconnectTest {
 
     private ISet<String> iSet;
+
+    @Override
+    String getServiceName() {
+        return SetService.SERVICE_NAME;
+    }
 
     @Override
     protected String addListener() {

--- a/hazelcast-client/src/test/java/com/hazelcast/client/listeners/TopicOnReconnectTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/listeners/TopicOnReconnectTest.java
@@ -22,6 +22,7 @@ import com.hazelcast.core.MessageListener;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
+import com.hazelcast.topic.impl.TopicService;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
@@ -30,6 +31,11 @@ import org.junit.runner.RunWith;
 public class TopicOnReconnectTest extends AbstractListenersOnReconnectTest {
 
     private ITopic<String> topic;
+
+    @Override
+    String getServiceName() {
+        return TopicService.SERVICE_NAME;
+    }
 
     @Override
     protected String addListener() {

